File: src/parsers/trading212parser.py
from src.models.brokers.trading212 import Trading212
from src.parsers.parser import Parser
#from ..parsers.parser import Parser
from src.models.transaction import Transaction
#from ..models.transaction import Transaction
from src.models.tax import Tax
from src.models.fee import Fee


import csv

class Trading212Parser(Parser):
    def parse(self, data):
        transactions = []
        with open(data, "r", encoding="utf-8-sig") as f:
            reader = csv.DictReader(f, delimiter=",")#csv.reader(f, delimiter=';')
            
            for row in reader:
                if(row["Time"] == ""):
                    continue
                
                date, time = row["Time"].split(" ")
                type = row["Action"]
                if("buy" in type.lower()):
                    type = "Buy"
                elif("sell" in type.lower()):
                    type = "Sell"
                elif("dividend" in type.lower()):
                    type = "Dividend"
                else:
                    continue 
                    
                ticker = row["Ticker"]
                isin = row["ISIN"]
                shares = float(row["No. of shares"])
                amount = float(row["Total"])
                amount_currency = row["Currency (Total)"]
                asset_currency = row["Currency (Price / share)"]
                
                # Obter fees
                fees = []
                # Obter todas as colunas que contenham "fee" no nome
                fee_names = [key for key in row.keys() if "fee" in key.lower()]
                for fee_name in fee_names:
                    try:
                        fee_amount = float(row[fee_name])
                    except ValueError:
                        continue
                    fee_currency = row[f"Currency ({fee_name})"]
                    fees.append(Fee(fee_name, fee_amount, fee_currency))
                    
                # Obter taxes
                taxes = []
                # Obter todas as colunas que contenham "tax" no nome
                tax_names = [key for key in row.keys() if "tax" in key.lower()]
                for tax_name in tax_names:
                    try:
                        tax_amount = float(row[tax_name])
                    except ValueError:
                        continue
                    tax_currency = row[f"Currency ({tax_name})"]
                    taxes.append(Tax(tax_name, tax_amount, tax_currency))
                
                transaction = Transaction(date, time, type, ticker, isin, shares, asset_currency, amount, amount_currency, taxes, fees, Trading212())
                if(transaction.type != ""):
                    transactions.append(transaction)

        return transactions
    
    
    

File: src/parsers/trading212parser.ts
import { Fee } from "../models/fee.js";
import { Tax } from "../models/tax.js";
import { Transaction } from "../models/transaction.js";
import { Trading212 } from "../models/brokers/trading212.js";
import { Parser } from "./parser.js";
import { Statement } from "../models/statement.js";
import { DateTime } from "luxon";

class Trading212Parser implements Parser {
  parse(fileData: string): Transaction[] {
    const transactions: Transaction[] = [];
    const data = fileData;
    const rows = data.split("\n").map((row) => row.split(","));
    const headers = rows.shift();

    if (!headers) {
      throw new Error("Invalid file data: no headers found");
    }

    rows.forEach((row) => {
      const record = Object.fromEntries(headers.map((h, i) => [h, row[i]]));
      if (!record["Time"]) return;

      const [date, time] = record["Time"].split(" ");
      const utcDate = DateTime.fromFormat(
        `${date} ${time}`,
        "yyyy-MM-dd HH:mm:ss", { zone: "utc" });

      let type = record["Action"].toLowerCase();
      if (type.includes("buy")) type = "Buy";
      else if (type.includes("sell")) type = "Sell";
      else if (type.includes("dividend")) type = "Dividend";
      else return;

      const ticker = record["Ticker"];
      const isin = record["ISIN"];
      const shares = parseFloat(record["No. of shares"]);
      const amount = parseFloat(record["Total"]);
      const amountCurrency = record["Currency (Total)"].replaceAll('"', "");
      const assetCurrency = record["Currency (Price / share)"];

      const fees = Object.keys(record)
        .filter((key) => key.toLowerCase().includes("fee"))
        .map((feeName) => {
          const feeAmount = parseFloat(record[feeName]);
          if (isNaN(feeAmount) || feeAmount < Number.EPSILON) return;
          const feeCurrency = record[`Currency (${feeName})`].replaceAll(
            '"',
            ""
          );
          return new Fee(feeName, feeAmount, feeCurrency);
        })
        .filter((fee) => fee != undefined);

      const taxes = Object.keys(record)
        .filter((key) => key.toLowerCase().includes("tax"))
        .map((taxName) => {
          const taxAmount = parseFloat(record[taxName]);
          if (isNaN(taxAmount) || taxAmount < Number.EPSILON) return;
          const taxCurrency = record[`Currency (${taxName})`].replaceAll(
            '"',
            ""
          );
          return new Tax(taxName, taxAmount, taxCurrency);
        })
        .filter((tax) => tax != undefined);

      let exchangeRate = 1 / parseFloat(record["Exchange rate"]);

      const transaction = new Transaction(
        utcDate,
        type,
        ticker,
        isin,
        shares,
        assetCurrency,
        amount,
        amountCurrency,
        new Trading212(),
        taxes,
        fees,
        exchangeRate
      );
      if (transaction.type) transactions.push(transaction);
    });

    // console.log("Transactions: " + JSON.stringify(transactions));

    return transactions;
  }
}

export { Trading212Parser };
